src/network/network_crypto.cpp

```diff
@@ -0,0 +1,455 @@
+	{
+		crypto_aead_write(&this->context, message.data(), mac.data(), nullptr, 0, message.data(), message.size());
+	}
+};
+
+/** Ensure the key does not get leaked when we're done with it. */
+X25519Key::~X25519Key()
+{
+	crypto_wipe(*this);
+}
+
+/**
+ * Create a new secret key that's filled with random bytes.
+ * @return The randomly filled key.
+ */
+/* static */ X25519SecretKey X25519SecretKey::CreateRandom()
+{
+	X25519SecretKey secret_key;
+	RandomBytesWithFallback(secret_key);
+	return secret_key;
+}
+
+/**
+ * Create the public key associated with this secret key.
+ * @return The public key.
+ */
+X25519PublicKey X25519SecretKey::CreatePublicKey() const
+{
+	X25519PublicKey public_key;
+	crypto_x25519_public_key(public_key.data(), this->data());
+	return public_key;
+}
+
+/**
+ * Create a new nonce that's filled with random bytes.
+ * @return The randomly filled nonce.
+ */
+/* static */ X25519Nonce X25519Nonce::CreateRandom()
+{
+	X25519Nonce nonce;
+	RandomBytesWithFallback(nonce);
+	return nonce;
+}
+
+/** Ensure the nonce does not get leaked when we're done with it. */
+X25519Nonce::~X25519Nonce()
+{
+	crypto_wipe(*this);
+}
+
+/**
+ * Create the handler, and generate the public keys accordingly.
+ * @param secret_key The secret key to use for this handler. Defaults to secure random data.
+ */
+X25519AuthenticationHandler::X25519AuthenticationHandler(const X25519SecretKey &secret_key) :
+	our_secret_key(secret_key), our_public_key(secret_key.CreatePublicKey()), nonce(X25519Nonce::CreateRandom())
+{
+}
+
+/* virtual */ void X25519AuthenticationHandler::SendRequest(Packet &p)
+{
+	p.Send_bytes(this->our_public_key);
+	p.Send_bytes(this->nonce);
+}
+
+/**
+ * Read the key exchange data from a \c Packet that came from the server,
+ * @param p The packet that has been received.
+ * @return True when the data seems correct.
+ */
+bool X25519AuthenticationHandler::ReceiveRequest(Packet &p)
+{
+	if (p.RemainingBytesToTransfer() != X25519_KEY_SIZE + X25519_NONCE_SIZE) {
+		Debug(net, 1, "[crypto] Received auth response of illegal size; authentication aborted.");
+		return false;
+	}
+
+	p.Recv_bytes(this->peer_public_key);
+	p.Recv_bytes(this->nonce);
+	return true;
+}
+
+/**
+ * Perform the key exchange, and when that is correct fill the \c Packet with the appropriate data.
+ * @param p The packet that has to be sent.
+ * @param derived_key_extra_payload The extra payload to pass to the key exchange.
+ * @return Whether the key exchange was successful or not.
+ */
+bool X25519AuthenticationHandler::SendResponse(Packet &p, std::string_view derived_key_extra_payload)
+{
+	if (!this->derived_keys.Exchange(this->peer_public_key, X25519KeyExchangeSide::CLIENT,
+			this->our_secret_key, this->our_public_key, derived_key_extra_payload)) {
+		Debug(net, 0, "[crypto] Peer sent an illegal public key; authentication aborted.");
+		return false;
+	}
+
+	X25519KeyExchangeMessage message{};
```

Comment:
The docs say it encrypts a few random bytes, but these are just zeros?

Reply:
Now it's filling it with random bytes. It must be something that did not survive the countless refactors that I've done over the time.